File: coding_skill_cpp/13_reference_counting.cpp
/* C++ Reference Counting - shared_ptr */
#include <iostream>
#include <memory>
using namespace std;

struct Buffer {
    Buffer() { cout << "[Buffer] 생성" << endl; }
    ~Buffer() { cout << "[Buffer] 소멸" << endl; }
    int data[100];
};

int main() {
    cout << "=== C++ Reference Counting ===" << endl;
    
    shared_ptr<Buffer> buf1 = make_shared<Buffer>();
    cout << "ref_count: " << buf1.use_count() << endl;
    
    {
        shared_ptr<Buffer> buf2 = buf1;
        cout << "ref_count: " << buf1.use_count() << endl;
    }
    
    cout << "ref_count: " << buf1.use_count() << endl;
    
    return 0;
}

File: coding_skill_py/13_reference_counting.py
"""Python Reference Counting - 내장 기능"""
import sys

class Buffer:
    def __init__(self):
        print("[Buffer] 생성")
    
    def __del__(self):
        print("[Buffer] 소멸")

def main():
    print("=== Python Reference Counting (내장) ===")
    
    buf1 = Buffer()
    print(f"ref_count: {sys.getrefcount(buf1) - 1}")
    
    buf2 = buf1
    print(f"ref_count: {sys.getrefcount(buf1) - 1}")
    
    del buf2
    print(f"ref_count: {sys.getrefcount(buf1) - 1}")

if __name__ == "__main__":
    main()
